### DISCUSSION TITLE
Questions about full waveform inversion of elastic waves using JUDI or Devito

### DISCUSSION THREAD

**Discussion by zhangxiaoshuotttt** [slimgroup/JUDI.jl#182] [General]
HI:

I have recently been experimenting with full-waveform elastic wave inversion using `JUDI` or `Devito`. I have been carefully reading your tutorials listed on Devito for the last week (for which I thank you for your hard work). 

Today I tried to perform full-waveform elastic wave inversion using Devtio or JUDI. I noticed that in this link, [Elastic](https://github.com/devitocodes/devito/tree/master/examples/seismic/elastic), there are no `AdjointOperator`, `GradientOperator`, or `BornOperator`, as in the [Acoustic]( https://github.com/devitocodes/devito/tree/master/examples/seismic/Acoustic) .

So the first question I want to ask is, if I want to try the full waveform inversion of the elastic wave, now these AdjoinToperators, GradienToperators have to be written by me, right?

Second, Because I want to read the code written by experts to learn more quickly, so I ask if you have these Operators related to the full waveform inversion of elastic waves at present. If not, I'll write it myself.

Finally, I would like to thank you again for the Tutorials you wrote in` Devito` and` JUDI`, they helped me a lot.

**Comment by mloubout** [comment]
Hi

Glad you enjoying it, and happy to provide guidance/help so you can get to where you want.

At the moment, as you pointed out, only forward elastic modeling is implemented in the devito examples. There is an open PR in devito to add the adjoint (and potential gradient/born for `vp`).

We would like to add this to JUDI as well at some point, time permitted, and any external contribution is greatly appreciated. You can also join the devito slack if you have additional questions or continue the discussion here.
